kong/concurrency.lua

```diff
@@ -109,6 +109,15 @@ function concurrency.with_coroutine_mutex(opts, fn)
     semaphore:post(1)
   end
 
+  -- to resolve deadlock issue in case the worker event thread
+  -- associated with `pcall(fn)` below got killed
+  -- and the `:post(1)` followed is skipped.
+  if semaphore:count() <= 0 then
```

Comment:
Restore missed comments:

<img width="463" alt="Screenshot 2024-05-13 at 15 12 39" src="https://github.com/Kong/kong/assets/6426329/e1bd561f-7171-4124-b5d0-5f9035e6cd6d">

Reply:
1. @samugi we should at least post 1 for unlock loop dependencies. Post 1 is safe, just as did here: https://github.com/Kong/kong/blob/master/kong/concurrency.lua#L101-L110.
2. @dndx the `pcall` failed has multiple causes. It runs within a timer thread spawn by worker events. If the thread aborts abnormally in the middle of `pcall`, there is no chance to `semaphore:post(1)`.

Reply:
Why would the thread abort in the worker event callback? Is this an intended behavior?

Reply:
@dndx for example, the worker 0 is OOM, then it is killed. If worker 0 is killed, then other workers cannot connect to it and error. Other workers will close the reconfigure thread when it is still applying new config (`pcall`).

Reply:
Again this has nothing to do with workers being killed. Coroutine mutexes don't have any meaning across process boundaries. @outsinre so please stop spreading this misinformation. I already answered in original PR that is exactly the same as here. And stop reopning this. It is just a bug that you are introducing.